tool/clean/clean_log_group/clean_log_group.go

```diff
@@ -0,0 +1,280 @@
+	creationTime := *logGroup.CreationTime
+
+	if creationTime >= times.creation {
+		return nil
+	}
+
+	lastLogTime := getLastLogEventTime(ctx, client, logGroupName)
+	if lastLogTime == 0 {
+		return nil
+	}
+
+	if lastLogTime < times.inactive {
+		logger.Printf("🚨 Worker: %d| Old & Inactive Log Group: %s (Created: %v, Last Event: %v)\n",
+			workerID, logGroupName, time.Unix(creationTime/1000, 0), time.Unix(lastLogTime/1000, 0))
+
+		mutex.Lock()
+		*logGroupsToDelete = append(*logGroupsToDelete, logGroupName)
+		mutex.Unlock()
+
+		if cfg.dryRun {
+			logger.Printf("🛑 Dry-Run: Would delete log group: %s", logGroupName)
+			return nil
+		}
+
+		return deleteLogGroup(ctx, client, logGroupName)
+	}
+
+	return nil
+}
+
+func deleteLogGroup(ctx context.Context, client *cloudwatchlogs.Client, logGroupName string) error {
+	_, err := client.DeleteLogGroup(ctx, &cloudwatchlogs.DeleteLogGroupInput{
+		LogGroupName: aws.String(logGroupName),
+	})
+	if err != nil {
+		return fmt.Errorf("deleting log group %s: %w", logGroupName, err)
+	}
+	logger.Printf("✅ Deleted log group: %s", logGroupName)
+	return nil
+}
+
+func fetchAndProcessLogGroups(ctx context.Context, client *cloudwatchlogs.Client,
+	logGroupChan chan<- *types.LogGroup, logGroupsToDelete *[]string, mutex *sync.Mutex) error {
+
+	var nextToken *string
+	describeCount := 0
+
+	for {
+		output, err := client.DescribeLogGroups(ctx, &cloudwatchlogs.DescribeLogGroupsInput{
+			NextToken: nextToken,
+		})
+		if err != nil {
+			return fmt.Errorf("describing log groups: %w", err)
+		}
+
+		logger.Printf("🔍 Described %d times | Found %d log groups\n", describeCount, len(output.LogGroups))
+
+		for _, logGroup := range output.LogGroups {
+			if isLogGroupException(*logGroup.LogGroupName) {
+				logger.Printf("⏭️ Skipping Log Group: %s (in exception list)\n", *logGroup.LogGroupName)
+				continue
+			}
+			logGroupChan <- &logGroup
+		}
+
+		if output.NextToken == nil {
+			break
+		}
+
+		mutex.Lock()
+		count := len(*logGroupsToDelete)
+		mutex.Unlock()
+
+		if count > cfg.deleteBatchCap {
+			break
+		}
+
+		nextToken = output.NextToken
+		describeCount++
+		logger.Printf("🔍 Processed %d log groups so far\n", count)
+	}
+
+	return nil
+}
+
+func getLastLogEventTime(ctx context.Context, client *cloudwatchlogs.Client, logGroupName string) int64 {
+	var latestTimestamp int64
+	var nextToken *string
+
+	for {
+		output, err := client.DescribeLogStreams(ctx, &cloudwatchlogs.DescribeLogStreamsInput{
+			LogGroupName: aws.String(logGroupName),
+			OrderBy:      types.OrderByLastEventTime,
+			Descending:   aws.Bool(true),
+			NextToken:    nextToken,
+		})
+		if err != nil {
+			logger.Printf("⚠️ Warning: Failed to retrieve log streams for %s: %v\n", logGroupName, err)
+			return 0
+		}
+
+		for _, stream := range output.LogStreams {
+			if stream.LastEventTimestamp != nil && *stream.LastEventTimestamp > latestTimestamp {
+				latestTimestamp = *stream.LastEventTimestamp
+			}
+		}
```

Comment:
If we're ordering it by last event time and it's descending, do we need to look beyond the first log stream?

Reply:
I am not 100% sure

Reply:
It is sorted. We should be able to just look at the first returned log stream.
https://github.com/aws/amazon-cloudwatch-agent/blob/9bdc4969b711f49312a3430a2611cfeb9aadb132/tool/clean/clean_log_group/clean_log_group.go#L260-L264